backend/hct_mis_api/one_time_scripts/program_cycle_data_migration.py

```diff
@@ -0,0 +1,175 @@
+
+
+def processing_with_active_program(payment_plans_list: list, default_cycle: ProgramCycle) -> None:
+    hhs_in_cycles_dict = dict()
+    for comparing_with_pp in payment_plans_list:
+        new_hh_ids = set(
+            [str(hh_id) for hh_id in comparing_with_pp.eligible_payments.values_list("household_id", flat=True)]
+        )
+        cycles = (
+            ProgramCycle.objects.filter(program_id=comparing_with_pp.program_id).exclude(id=default_cycle.id).only("id")
+        )
+        for cycle in cycles:
+            cycle_id_str = str(cycle.id)
+            if cycle_id_str not in hhs_in_cycles_dict:
+                hhs_in_cycles_dict[cycle_id_str] = set(
+                    [
+                        str(hh_id)
+                        for hh_id in Payment.objects.filter(parent__program_cycle=cycle).values_list(
+                            "household_id", flat=True
+                        )
+                    ]
+                )
+            hh_ids_in_cycles = hhs_in_cycles_dict[cycle_id_str]
+
+            # check any conflicts
+            if new_hh_ids.intersection(hh_ids_in_cycles):
+                continue
+
+            TargetPopulation.objects.filter(id=comparing_with_pp.target_population_id).update(program_cycle=cycle)
+            comparing_with_pp.program_cycle = cycle
+            hhs_in_cycles_dict[cycle_id_str].update(new_hh_ids)
+            break
+
+        if not comparing_with_pp.program_cycle:
+            cycle = create_new_program_cycle(
+                str(comparing_with_pp.program_id),
+                ProgramCycle.ACTIVE,
+                comparing_with_pp.start_date.date(),
+                comparing_with_pp.end_date.date(),
+            )
+            TargetPopulation.objects.filter(id=comparing_with_pp.target_population_id).update(program_cycle=cycle)
+            comparing_with_pp.program_cycle = cycle
+            hhs_in_cycles_dict[str(cycle.id)] = new_hh_ids
+
+        comparing_with_pp.save(update_fields=["program_cycle"])
+
+
+def program_cycle_data_migration() -> None:
+    start_time = timezone.now()
+    logger.info("Hi There! Started Program Cycle Data Migration.")
+    logger.info(f"Cycles before running creation: {ProgramCycle.objects.all().count()}")
+    for ba in BusinessArea.objects.all().only("id", "name"):
+        logger.info(f"Started processing {ba.name}...")
+
+        # FINISHED program
+        for finished_program in (
+            Program.objects.filter(business_area_id=ba.id)
+            .exclude(status__in=[Program.DRAFT, Program.ACTIVE])
+            .only("id", "name", "start_date", "end_date", "status")
+        ):
+            start_data = finished_program.start_date
+            end_data = finished_program.end_date
+            for pp in PaymentPlan.objects.filter(program_id=finished_program.id).only(
+                "id", "program_id", "target_population_id"
+            ):
+                processing_with_finished_program(pp, start_data, end_data)
```

Comment:
This is still called per PPlan but it is not necessary. As far as I understand, we can make this function per Program, create a ProgramCycle there and just assign it to all PPlans and TPs from this Program at once.
There is no need to update it in a PPlan loop, since the data is taken from their Program anyway.